File: src/main/java/org/warp/picalculator/math/rules/FractionsRule1.java
package org.warp.picalculator.math.rules;

import org.warp.picalculator.Error;
import org.warp.picalculator.math.Function;
import org.warp.picalculator.math.MathContext;
import org.warp.picalculator.math.functions.Division;
import org.warp.picalculator.math.functions.Number;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;

/**
 * Fractions rule<br>
 * <b>0 / a = 0</b>
 * 
 * @author Andrea Cavalli
 *
 */
public class FractionsRule1 {

	public static boolean compare(Function f) {
		final MathContext root = f.getMathContext();
		final Division fnc = (Division) f;
		if (fnc.getParameter1() instanceof Number) {
			final Number numb1 = (Number) fnc.getParameter1();
			if (numb1.equals(new Number(root, 0))) {
				if (fnc.getParameter2() instanceof Number) {
					final Number numb2 = (Number) fnc.getParameter2();
					if (numb2.equals(new Number(root, 0))) {
						return false;
					}
				}
				return true;
			}
		}
		return false;
	}

	public static ObjectArrayList<Function> execute(Function f) throws Error {
		final ObjectArrayList<Function> result = new ObjectArrayList<>();
		result.add(new Number(f.getMathContext(), 0));
		return result;
	}

}

File: src/main/resources/rules/FractionsRule1.js
//Imports


var Error = org.warp.picalculator.Error;
var Function = org.warp.picalculator.math.Function;
var MathContext = org.warp.picalculator.math.MathContext;
var Division = org.warp.picalculator.math.functions.Division;
var Number = org.warp.picalculator.math.functions.Number;

var ObjectArrayList = Java.type("it.unimi.dsi.fastutil.objects.ObjectArrayList");

/**
 * Fractions rule
 * 0 / a = 0
 * 
 * @author Andrea Cavalli
 *
 */
var rule = {
	// Rule name
	getRuleName: function() {
		return "FractionsRule1";
	},
	// Rule type
	getRuleType: function() {
		return RuleType.CALCULATION;
	},
	/* Rule function
	   Returns:
	     - null if it's not executable on the function "f"
	     - An ObjectArrayList<Function> if it did something
	*/

	execute: function(f) {
		var isExecutable;
		var root = f.getMathContext();
		if (ScriptUtils.instanceOf(f, Division.class)) {
			var fnc = f;
			if (ScriptUtils.instanceOf(fnc.getParameter1(), Number.class)) {
				var numb1 = fnc.getParameter1();
				if (numb1.equals(new Number(root, 0))) {
					if (ScriptUtils.instanceOf(fnc.getParameter2(), Number.class)) {
						var numb2 = fnc.getParameter2();
						if (numb2.equals(new Number(root, 0)) == false) {
							isExecutable = true;
						}
					} else {
						isExecutable = true;
					}
				}
			}
		}
	
		if (isExecutable) {
			var result = new ObjectArrayList();
			result.add(new Number(f.getMathContext(), 0));
			return result;
		} else {
			return null;
		}
	}
}

//Add this rule to the list of rules
RulesManager.addRule(engine.getInterface(rule, Rule.class));